lib/using_restricted_array.rb

```diff
@@ -6,56 +6,117 @@
+    i += 1
+  end
+  print "#{array[i]}"
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) - linear (will iterate n times, until it finds the value searched)
+# Space complexity: O(1) - constant (won't change regardless of array size)
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  i = 0
+  while i < length
+    if value_to_find == array[i]
+      return true
+    else
+      i += 1
+    end
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: for n = array.length, time complexity is linear O(n)
+# Space complexity: O(1) - constant
 def find_largest(array, length)
-  raise NotImplementedError
+  i = 0
+  largest = 0
+  while i < length
+    if array[i] > largest
+      largest = array[i]
+      i += 1
+    else
+      i += 1
+    end
+  end
+  return largest
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear O(n) - for n = array.length, will iterate n times to check smallest value
+# Space complexity: O(1) - constant (remains the same regardless of array size)
 def find_smallest(array, length)
-  raise NotImplementedError
+  i = 0
+  smallest = array[0]
+  while i < length
+    if array[i] < smallest
+      smallest = array[i]
+    end
+    i += 1
+  end
+  return smallest
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) - for n = array.length, will iterate n times to reverse all values
+# Space complexity: O(1) - constant
 def reverse(array, length)
-  raise NotImplementedError
+  original_index = 0
+  swap_index = length - 1
+
+  while original_index < swap_index
+    original_value = array[original_index]
+    swap_value = array[swap_index]
+
+    array[original_index] = swap_value
+    array[swap_index] = original_value
+
+    original_index += 1
+    swap_index -= 1
+  end
+  return array
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: it depends; if best case applies, then linear O(1). If not, log(n)
+# Space complexity: constant O(1) - (remains the same)
 def binary_search(array, length, value_to_find)
-  raise NotImplementedError
+  start = 0
+  finish = length - 1
+
+  while finish > start
+    middle = (start + finish) / 2
+    if array[middle] == value_to_find
+      return true
+    elsif array[middle] < value_to_find
+      start = middle + 1
+    elsif array[middle] > value_to_find
+      start = middle - 1
+    end
+  end
+  return array[start] == value_to_find ? true : false
```

Comment:
Just a bit of boolean zen.

```suggestion
  return array[start] == value_to_find
```